Change handling of school addresses from the search tool

erikphansen opened the pull request:
## Description
- [x] Update package.json to use the new [version of JSON schema](https://github.com/department-of-veterans-affairs/vets-json-schema/pull/254) when it's ready
- ~Combine the second and third address fields so no data is lost; this
is especially important for some international addresses that the search
tool returns as the many of those contain important address info in the
`address3` field.~ We have instead moved to adding a `street3` field to the address object we send to the backend.
- If the address from the search tool is domestic, transform the country
to the prop enum value, "United States".
- ~If the address is not domestic, send the country as-is and add the
`viaSearchTool=true` field on the `address` so the backend will use the
school address schema that accepts any string for the country value.~ We are instead simply adding a less stringent version of the international address schema that allows for any string for the country, rather than requiring the country to be selected from the defined list.

## Testing done
Updated unit tests to test handling of domestic and international addresses from the search tool.

## Testing Plan
The format of addresses that come from the search tool vary widely. This calls for manual testing, especially of international addresses, to assure that the correct address data gets passed to the backend.

## Screenshots
N/A

## Acceptance Criteria (Definition of Done)

#### Unique to this PR
- [x] Use the stricter schema for domestic addresses, use the more forgiving address schema for international addresses.

#### Applies to all PRs

- [ ] Appropriate logging
- [ ] Documentation has been updated, if applicable
- [x] Provide link to originating GitHub issue, or connected to it via ZenHub
- [x] Does not contain any sensitive information (i.e. PII/credentials/internal URLs/etc., in logging, hardcoded, or in specs)
- [ ] Provide which alerts would indicate a problem with this functionality (if applicable)

erikphansen committed:
Change handling of school addresses from the search tool

- Combine the second and third address fields so no data is lost; this
is especially important for some international addresses that the search
tool returns as the many of those contain important address info in the
`address3` field.
- If the address from the search tool is domestic, transform the country
to the prop enum value, "United States".
- If the address is not domestic, send the country as-is and add the
`viaSearchTool=true` field on the `address` so the backend will use the
school address schema that accepts any string for the country value.

rianfowler commented on src/applications/edu-benefits/components/SchoolSelectField.jsx:
```diff
@@ -84,21 +84,27 @@ export class SchoolSelectField extends React.Component {
-        state,
-        postalCode: zip,
-      }
+      address
```
Did we move away from submitting the facility code for these addresses?

erikphansen replied:
Yes. Lihan requested that the first day both you and Claire were gone. Since the search tool provided all of the required address info to the frontend it seemed doable.

rianfowler replied:
We really shouldn't be doing these kinds of data transformations in the client. If we can't submit exactly what we get back from the API then something should change on the server or the data.

cehsu replied:
Fascinating point @rianfowler, looping @jholton for any additional thoughts from the backend perspective on this idea for organizing logic.

cehsu commented on src/applications/edu-benefits/tests/components/SchoolSelectField.unit.spec.jsx:
```diff
@@ -291,46 +291,90 @@ describe('<SchoolSelectField>', () => {
       name: 'testName',
-      state: 'testState'
+      state: 'testState',
+      zip: '12345',
```
Do international institution addresses use the zip and state fields?

erikphansen replied:
I'm not sure and given that the international address formats we get back from the search tool are a little weird I don't think we can count on them always missing them. @rianfowler likely knows more about the different address formats we get from the search tool.

cehsu replied:
is nothing sacred?

cehsu commented on src/applications/edu-benefits/components/SchoolSelectField.jsx:
```diff
@@ -84,21 +84,27 @@ export class SchoolSelectField extends React.Component {
+    const isDomesticAddress = country === 'USA';
+    let street2 = address2;
+    if (address3) {
+      street2 = `${street2} ${address3}`;
```
We might want to have the backend handle this as they're already performing other schema related translations.

erikphansen replied:
Absolutely, if we add a `street3` to the schema.

cehsu reviewed (commented):
Looks good so far. I had a few comments/question for your consideration.

erikphansen committed:
Use the `street3` field

rianfowler commented on src/applications/edu-benefits/components/SchoolSelectField.jsx:
```diff
@@ -84,21 +84,24 @@ export class SchoolSelectField extends React.Component {
-    const backendFriendlyCountries = {
-      USA: 'United States'
+    const isDomesticAddress = country === 'USA';
+    const address = {
```
Maybe put this transformation into a helper.

erikphansen replied:
Nice. Will do.

rianfowler commented on src/applications/edu-benefits/components/SchoolSelectField.jsx:
```diff
@@ -84,21 +84,24 @@ export class SchoolSelectField extends React.Component {
 
   handleOptionClick = ({ address1, address2, address3, city, facilityCode, name, state, zip, country }) => {
```
Alphabetize this argument list.

erikphansen replied:
Ha, should I admit I was looking at the order of this list and thought it wasn't a logical way to order address info? Alphabetical didn't even enter my mind.

erikphansen committed:
Refactor address transforms to a helper

erikphansen committed:
Bypass the JSON stringify transformer

The default transformer was completely removing address objects because
it was very aggressively removing address objects that it deemed
invalid. In this case that meant if the address did not contain a zip
code/postal code, the entire address would be removed from the school
object

erikphansen commented on src/applications/edu-benefits/feedback-tool/helpers.js:
```diff
@@ -26,7 +26,7 @@ export function fetchInstitutions({ institutionQuery, page, onDone, onError }) {
 
 export function transform(formConfig, form) {
-  const formData = transformForSubmit(formConfig, form);
+  const formData = transformForSubmit(formConfig, form, null);
```
This will make the `transformForSubmit()` method no longer use the very aggression JSON stringify transformer that's removing international addresses that lack a zip/postal code (which included all int'l addresses added via the school search tool).

erikphansen replied:
the transformForSubmit method is defined [here](https://github.com/usds/us-forms-system/blob/master/lib/js/helpers.js#L382)
the stringify transformer I'm avoiding using is defined [here](https://github.com/usds/us-forms-system/blob/master/lib/js/helpers.js#L336)

cehsu replied:
Would it make sense to create an issue on the us-forms-system repo to break the cleaning logic out of the stringify transformer?

erikphansen replied:
yeah, I talked to Anne about this late yesterday and she agrees that the problematic JSON stringify transformer code is ooooold and the transform logic should be refactored. We're not sure the best approach, but we should document this.

cehsu commented on src/applications/edu-benefits/feedback-tool/config/form.js:
```diff
@@ -383,6 +383,9 @@ const formConfig = {
                     street2: {
                       'ui:title': 'Address line 2'
                     },
+                    street3: {
```
We may need to transform this on the FE until the backend has included this transformation in its transformation list.

cehsu replied:
Nevermind, it looks like this work is already done. Way to go @jholton.
https://github.com/department-of-veterans-affairs/vets-api/blob/f367cbde99b9fc812d6bd29bab06b0b312a69ea2/app/models/gi_bill_feedback.rb

erikphansen replied:
@cehsu I only see that transform happening [here](https://github.com/department-of-veterans-affairs/vets-api/blob/f367cbde99b9fc812d6bd29bab06b0b312a69ea2/app/models/gi_bill_feedback.rb#L34), which is no longer gonna run.

erikphansen replied:
Since the backend needs to be updated before we can release this I don't think we should bandaid it on the frontend.

cehsu replied:
Should we create a backend ticket to update the international school address transformation so that street3 is included when condensing international addresses onto only the street1 and street2 fields?

erikphansen replied:
ticket: https://app.zenhub.com/workspace/o/department-of-veterans-affairs/vets.gov-team/issues/13074

cehsu reviewed (approved):
LGTM. IIRC salesforce doesn't support street3, so we may need to temporarily combine street2 and street3 on the frontend (as you previously had set up), until the backend includes this transformation in its list.

erikphansen committed:
prevent sending search tool address fields that are too long

erikphansen committed:
Remove `viaSearchTool`, upgrade vets-json-schema

cehsu commented on src/applications/edu-benefits/feedback-tool/config/form.js:
```diff
@@ -35,10 +35,10 @@ const {
 const { address: schoolAddress, name: schoolName } = school.properties;
-const domesticSchoolAddress = schoolAddress.oneOf[0];
-const internationalSchoolAddress = schoolAddress.oneOf[1];
+const domesticSchoolAddress = schoolAddress.anyOf[0];
```
Very well done 👏 , great catch using anyOf instead of oneOf.

erikphansen replied:
This caused so much confusion this morning. Then things still didn't work when I switch to `anyOf` because of a stupid copy/paste error I'd introduced in my code :(

cehsu commented on src/applications/edu-benefits/feedback-tool/helpers.js:
```diff
@@ -167,3 +178,61 @@ const applicantRelationshipEventMap = {
+ * school.address schema.
+ */
+export function transformSearchToolAddress({ address1, address2, address3, city, country, state, zip }) {
+  const isDomesticAddress = country === 'USA';
```
It seems inconsistent to use an abbreviation for USA only in the school address country enum list, would it make sense to create a ticket to convert this enum to the full name? If so, we could use the country value below on line 195 rather than a magic string.

erikphansen replied:
you might be misreading this (or I'm misunderstanding)? the search tool returns 'USA' and we're transforming it to 'United States'

cehsu replied:
Thanks for clarifying, that makes sense.

cehsu commented on src/applications/edu-benefits/feedback-tool/helpers.js:
```diff
@@ -167,3 +178,61 @@ const applicantRelationshipEventMap = {
+  if (isDomesticAddress) {
+    return {
+      country: 'United States',
+      street: typeof address1 === 'string'
```
Do we anticipate that these might not be strings? In what situation would that be the case?

erikphansen replied:
no. but they might be null or maybe undefined. and we don't want to `slice()` on something we can't slice.

cehsu replied:
That makes sense, if it exists it will be a string, so we can simply could do something like:
```
street: address1 && address.slice(....)
```

But what you've got works great.

erikphansen replied:
Ha, yours is totally better. Thanks for catching this!

cehsu reviewed (approved):
LGTM. I had a few questions and comments for your consideration.

erikphansen committed:
refactor string trimming
